ext/standard/basic_functions.c

```diff
@@ -1582,9 +1582,12 @@ static void fci_release(zend_fcall_info *fci, zend_fcall_info_cache *fci_cache)
 void user_shutdown_function_dtor(zval *zv) /* {{{ */
 {
 	php_shutdown_function_entry *shutdown_function_entry = Z_PTR_P(zv);
+	zend_fcall_info *fci = &shutdown_function_entry->fci;
 
-	zend_fcall_info_args_clear(&shutdown_function_entry->fci, true);
-	fci_release(&shutdown_function_entry->fci, &shutdown_function_entry->fci_cache);
+	if (fci) {
+		zend_fcall_info_args_clear(fci, true);
+		fci_release(fci, &shutdown_function_entry->fci_cache);
+	}
```

Comment:
This seems to point to a bug _elsewhere_ it should not be possible to not have an FCI here.

Reply:
Furthermore, the patch also doesn't really make sense: fci is address-taken from a struct member of shutdown_function_entry, so it can only be NULL if shutdown_function_entry is NULL, but in that case it's undefined behaviour because you aren't allowed to do pointer math on NULL...
I think we should all run our development tools with ASAN + UBSAN enabled.

Reply:
Agreed that we should check `shutdown_function_entry` here, and not `shutdown_function_entry->fci`.

My understanding of the issue is that we OOM here when registering a new shutdown function:
https://github.com/php/php-src/blob/909d331d0a568efa576bdf04911dd69ef6b00fa8/Zend/zend_hash.h#L858 
as a consequence, `Z_PTR_P(zv)` remains `NULL` (it is initialized to `NULL` above), so `shutdown_function_entry` is `NULL` here: https://github.com/php/php-src/blob/909d331d0a568efa576bdf04911dd69ef6b00fa8/ext/standard/basic_functions.c#L1584

So the root cause is `zend_hash_next_index_insert_mem()`. We should probably change this function so that it doesn't leave NULL entries in case of allocation failure. We could move the malloc/memcpy before the insert, initialize tmp to that, and free when insert fails. WDYT?

Reply:
I think moving the allocation before the insert is the right approach.

Reply:
Thanks all for your feedback, just trying to take of "day to day" stuff as much as I can :) but at least I get better understanding of the internals.